Carbon.Windows.Installer/Functions/Get-InstalledProgram.ps1

```diff
@@ -0,0 +1,241 @@
+
+function Get-InstalledProgram
+{
+    <#
+    .SYNOPSIS
+    Gets information about the programs installed on the computer.
+    
+    .DESCRIPTION
+    The `Get-CInstalledProgram` function is the PowerShell equivalent of the Programs and Features/Apps and Features
+    settings UI. It inspects the registry to determine what programs are installed. When running as an administrator, it
+    returns programs installed for *all* users, not just the current user.
+
+    The function looks in the following registry keys for install information:
+
+    * HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Uninstall
+    * HKEY_LOCAL_MACHINE\Software\Wow6432Node\Microsoft\Windows\CurrentVersion\Uninstall
+    * HKEY_USERS\*\Software\Microsoft\Windows\CurrentVersion\Uninstall\*'
+
+    A key is skipped if:
+
+    * it doesn't have a `DisplayName` value.
+    * it has a `ParentKeyName` value.
+    * it has a `SystemComponent` value and its value is `1`.
+
+    `Get-CInstalledProgram` tries its best to get accurate data. The following properties either aren't stored
+    consistently, is in strange formats, can't be parsed, etc.
+
+    * The `ProductCode` property is set to `[Guid]::Empty` if the software doesn't have a product code.
+    * The `User` property will only be set for software installed for specific users. For global software, the `User`
+      property will be `[String]::Empty`.
+    * The `InstallDate` property is set to `[DateTime]::MinValue` if the install date can't be determined.
+    * The `Version` property is `$null` if the version can't be parsed.
+    
+    .EXAMPLE
+    Get-CInstalledProgram | Sort-Object 'DisplayName'
+
+    Demonstrates how to get a list of all the installed programs, similar to what the Programs and Features settings UI
+    shows. The returned objects are not sorted, so you'll usually want to pipe the output to `Sort-Object`.

```

Comment:
Seems like the default should be to sort by DisplayName like the "Programs and Features" applet.

Reply:
This function already takes a little over 2 seconds:

```
$ 1..10 | % { measure-command { Get-CInstalledProgram } } | ft Minutes,Seconds,Milliseconds

Minutes Seconds Milliseconds
------- ------- ------------
      0       2          166
      0       2          184
      0       2          186
      0       2          249
      0       2          586
      0       2          239
      0       2          272
      0       2          199
      0       2          187
      0       2          170


```

If I add a sort, it looks like it takes an extra quarter to half a second:

```
Minutes Seconds Milliseconds
------- ------- ------------
      0       2          841
      0       2          462
      0       2          573
      0       2          822
      0       2          499
      0       2          335
      0       2          163
      0       2          170
      0       2          607
      0       2          529
```

Reply:
Yikes. Definitely don't want to pre-sort then.

Reply:
I spent a little time optimizing this function, and got the timings down almost 50%, including the sort (which adds about 1/4 second).

```
$ % { measure-command { Get-CInstalledProgram } } | ft Minutes,Seconds,Milliseconds

Minutes Seconds Milliseconds
------- ------- ------------
      0       1          906
      0       1          134
      0       1          191
      0       1          167
      0       1          170
      0       1          129
      0       1           94
      0       1          100
      0       1          150
      0       1          134


```